staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding.go

```diff
@@ -0,0 +1,115 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package shufflesharding
+
+import (
+	"fmt"
+	"math/big"
+	"sync"
+)
+
+// MaxHashBits is the max bit length which can be used from hash value.
+// If we use all bits of hash value, the critical(last) card shuffled by
+// Dealer will be uneven to 2:3 (first half:second half) at most,
+// in order to reduce this unevenness to 32:33, we set MaxHashBits to 60 here.
+const MaxHashBits = 60
+
+// RequiredEntropyBits will figure out the required entropy bits from the given
+// deckSize and handSize by calculating bits of (deckSize^handSize).
+// RequiredEntropyBits must be less or equal than MaxHashBits
+// if we want to create and use Dealer.
+func RequiredEntropyBits(deckSize, handSize int) int {
+	return big.NewInt(1).Exp(big.NewInt(int64(deckSize)), big.NewInt(int64(handSize)), nil).BitLen()
+}
+
+// Dealer contains some necessary parameters and provides some methods for shuffle sharding.
+// Dealer is thread-safe.
+type Dealer struct {
+	deckSize int
+	handSize int
+
+	// rawCardsPool is a pool for caching and reusing rawCards slices to improve performance
+	rawCardsPool *sync.Pool
+}
+
+// NewDealer will create a Dealer with the given deckSize and handSize, will return error when
+// deckSize or handSize is invalid as below.
+// 1. deckSize or handSize is not positive
+// 2. handSize is greater than deckSize
+// 3. required entropy bits of deckSize and handSize is greater than MaxHashBits
+func NewDealer(deckSize, handSize int) (*Dealer, error) {
+	if deckSize <= 0 || handSize <= 0 {
+		return nil, fmt.Errorf("deckSize %d or handSize %d is not positive", deckSize, handSize)
+	}
+	if handSize > deckSize {
+		return nil, fmt.Errorf("handSize %d is greater than deckSize %d", handSize, deckSize)
+	}
+	if RequiredEntropyBits(deckSize, handSize) > MaxHashBits {
+		return nil, fmt.Errorf("required entropy bits of deckSize %d and handSize %d is greater than %d", deckSize, handSize, MaxHashBits)
+	}
+
+	return &Dealer{
+		deckSize: deckSize,
+		handSize: handSize,
+		rawCardsPool: &sync.Pool{
+			New: func() interface{} {
+				return make([]int, handSize)
+			},
+		},
+	}, nil
+}
+
+// Deal shuffles a card deck of the given size and deals a hand of the given size,
+// using the given hashValue as a source of entropy.
+// This function makes handSize sequential calls to pick, one for each dealt card.
+// Each card is identified by an integer in the range [0, deckSize).
+// For example, for deckSize=128 and handSize=4 this function might call pick(14); pick(73); pick(119); pick(26).
+// This function can be extended with different pick functions.
+func (d *Dealer) Deal(hashValue uint64, pick func(int)) {
+	rawCards := d.rawCardsPool.Get().([]int)
+	defer d.rawCardsPool.Put(rawCards)
```

Comment:
There is a much simpler and more efficient way to do this:
```
    var rawCards [15]int
```

Reply:
+1, I would not use the sync.Pool for this.

Reply:
@MikeSpreitzer @lavalamp  I'm sorry I don't think I've got your points, sync.Pool will get us out of repeated allocation for rawCards slice in different threads... Why allocating rawCards slice every time is a simper and more efficient way?

Reply:
I am suggesting a fixed-size array, not a slice.  The fixed-size array will be allocated on the heap.

Reply:
Also, BTW, I think "remainders" was a better than for this variable than "rawCards".

Reply:
Got your point, will update soon ~